clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,216 @@
+    // expected-note@+1{{in instantiation of function template specialization}}
+    kernel_single_task<class kernel9>(l10);
+  }
+
+  // kernel6 - expect no error
+  // Make a lambda
+  // Make another lambda that captures the first one
+  // Call the builtin on the first lambda (in a constexpr context)
+  // Pass the second lambda to a kernel
+  // Test that passing a lambda to the unique stable name builtin and then
+  // passing a second lambda capturing the first one to the kernel does not
+  // throw an error because the first lambda is not involved in naming the
+  // kernel i.e., the mangling does not change
+  auto l5 = []() { return 1; };
+  auto l6 = [l5]() { return 2; };
+  constexpr const char *l5_output = __builtin_unique_stable_name(l5);
+  kernel_single_task<class kernel6>(l6); // #kernel6call
+
+  // kernel11 - expect an error
+  // Make a lambda
+  // Call the builtin on the first lambda (in a constexpr context)
+  // Pass the lambda as a template template parameter to a kernel
+  // Test that passing a lambda to the unique stable name builtin and then
+  // passing it to the kernel as a template template parameter causes a
+  // diagnostic on the kernel use due to the change in results to the stable
+  // name
+  auto l11 = []() { return 1; };
+  constexpr const char *l11_output = __builtin_unique_stable_name(l11); // #USN_l11
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_l11{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel11>(S<Tangerine, decltype(l11)>{});
+
+  // kernel12 - expect an error
+  // Test that passing a lambda to the unique stable name builtin within a macro
+  // and then calling the macro within the kernel causes an error on the kernel
+  // due to the change in the results of the stable name
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_MACRO12{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel12>(
+      []() {
+        MACRO12(); // #USN_MACRO12
+      });
+}
+
+namespace NS {}
+
+void f() {
+  // expected-error@+1{{use of undeclared identifier 'bad_var'}}
+  __builtin_unique_stable_name(bad_var);
+  // expected-error@+1{{use of undeclared identifier 'bad'}}
+  __builtin_unique_stable_name(bad::type);
+  // expected-error@+1{{no member named 'still_bad' in namespace 'NS'}}
+  __builtin_unique_stable_name(NS::still_bad);
+
+  // FIXME: warning expected but none emitted
+  int i = 0;
+  __builtin_unique_stable_name(i++);
+
+  int j = 55;
+  __builtin_unique_stable_name(int[++j]); // no warning expected
```

Comment:
```suggestion
  // Tests that use within a VLA does not diagnose as a side-effecting use in an
  // unevaluated context because the use within a VLA extent forces evaluation.
  int j = 55;
  __builtin_unique_stable_name(int[++j]); // no warning expected
```